File: web/src/types/telemetry.ts
import { v4 as uuidv4 } from 'uuid';

export interface TelemetryData {
  id: string;
  truck_id: string;
  timestamp: string;
  location: [number, number]; // [longitude, latitude]
  speed_kmh: number;
  heading: number;
  sensors: SensorData;
  cameras: CameraData | null;
  scenario: string | null;
  created_at: string;
}

export interface SensorData {
  gps: GpsData;
  obd: ObdData;
  imu: ImuData;
  tpms: TpmsData;
}

export interface GpsData {
  latitude: number;
  longitude: number;
  altitude: number;
  speed_kmh: number;
  heading: number;
  satellites: number;
  fix_quality: number;
}

export interface ObdData {
  rpm: number;
  speed_kmh: number;
  coolant_temp: number;
  fuel_level: number;
  engine_load: number;
  throttle_pos: number;
}

export interface ImuData {
  accel_x: number;
  accel_y: number;
  accel_z: number;
  gyro_x: number;
  gyro_y: number;
  gyro_z: number;
}

export interface TpmsData {
  front_left: TireSensor;
  front_right: TireSensor;
  rear_left: TireSensor;
  rear_right: TireSensor;
}

export interface TireSensor {
  pressure_psi: number;
  temperature_c: number;
  battery_percent: number;
  alert: boolean;
}

export interface CameraData {
  front_camera: CameraFrameRef | null;
  driver_camera: CameraFrameRef | null;
  cargo_camera: CameraFrameRef | null;
}

export interface CameraFrameRef {
  frame_id: string;
  timestamp: string;
  url: string;
  thumbnail_url: string | null;
  width: number;
  height: number;
  format: string;
  size_bytes: number;
  is_keyframe: boolean;
  meta :FrameMetadata;
}

export interface FrameMetadata {
  exposure_us: number | null;
  gain_db: number | null;
  temperature_c: number | null;
  gps_lat: number | null;
  gps_lon: number | null;
  speed_kmh: number | null;
}

export interface TelemetrySummary {
  truck_id: string;
  last_timestamp: string;
  last_location: [number, number];
  last_speed_kmh: number;
  last_heading: number;
  last_rpm: number;
  last_coolant_temp: number;
  last_fuel_level: number;
  last_accel_x: number;
  last_accel_y: number;
  last_accel_z: number;
  tire_pressures: [number, number, number, number];
  tire_temperatures: [number, number, number, number];
}

export interface TelemetryStatistics {
  truck_id: string;
  start_time: string;
  end_time: string;
  avg_speed_kmh: number;
  max_speed_kmh: number;
  min_speed_kmh: number;
  avg_rpm: number;
  max_rpm: number;
  min_rpm: number;
  avg_coolant_temp: number;
  max_coolant_temp: number;
  min_coolant_temp: number;
  avg_fuel_level: number;
  min_fuel_level: number;
  total_distance_km: number;
  total_fuel_consumed_liters: number;
  harsh_braking_events: number;
  rapid_acceleration_events: number;
  overspeeding_events: number;
}

File: truck-central-server/src/models/telemetry.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};
use geo::Point;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelemetryData {
    pub id: Uuid,
    pub truck_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub location: Point<f64>,
    pub speed_kmh: f32,
    pub heading: f32,
    pub sensors: SensorData,
    pub cameras: Option<CameraData>,
    pub scenario: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SensorData {
    pub gps: GpsData,
    pub obd: ObdData,
    pub imu: ImuData,
    pub tpms: TpmsData,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GpsData {
    pub latitude: f64,
    pub longitude: f64,
    pub altitude: f32,
    pub speed_kmh: f32,
    pub heading: f32,
    pub satellites: u8,
    pub fix_quality: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObdData {
    pub rpm: u16,
    pub speed_kmh: u8,
    pub coolant_temp: i8,
    pub fuel_level: u8,
    pub engine_load: u8,
    pub throttle_pos: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImuData {
    pub accel_x: f32,
    pub accel_y: f32,
    pub accel_z: f32,
    pub gyro_x: f32,
    pub gyro_y: f32,
    pub gyro_z: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TpmsData {
    pub front_left: TireSensor,
    pub front_right: TireSensor,
    pub rear_left: TireSensor,
    pub rear_right: TireSensor,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TireSensor {
    pub pressure_psi: f32,
    pub temperature_c: f32,
    pub battery_percent: u8,
    pub alert: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CameraData {
    pub front_camera: Option<CameraFrameRef>,
    pub driver_camera: Option<CameraFrameRef>,
    pub cargo_camera: Option<CameraFrameRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CameraFrameRef {
    pub frame_id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub url: String,
    pub thumbnail_url: Option<String>,
    pub width: u32,
    pub height: u32,
    pub format: String,
    pub size_bytes: u64,
    pub is_keyframe: bool,
    pub meta FrameMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FrameMetadata {
    pub exposure_us: Option<u32>,
    pub gain_db: Option<f32>,
    pub temperature_c: Option<f32>,
    pub gps_lat: Option<f64>,
    pub gps_lon: Option<f64>,
    pub speed_kmh: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TelemetrySummary {
    pub truck_id: Uuid,
    pub last_timestamp: DateTime<Utc>,
    pub last_location: Point<f64>,
    pub last_speed_kmh: f32,
    pub last_heading: f32,
    pub last_rpm: u16,
    pub last_coolant_temp: i8,
    pub last_fuel_level: u8,
    pub last_accel_x: f32,
    pub last_accel_y: f32,
    pub last_accel_z: f32,
    pub tire_pressures: [f32; 4],
    pub tire_temperatures: [f32; 4],
}